control-plane/agents/src/bin/core/volume/scheduling.rs

```diff
@@ -143,3 +171,95 @@ pub(crate) async fn resizeable_replicas(
+                        }
+                        !matching_key && !label.affinity.is_empty()
+                    }
+
+                    NodeTopology::Explicit(_) => false,
+                },
+                None => false,
+            };
+            let pool_affinity = match &topology.pool {
+                Some(pool_topology) => match pool_topology {
+                    PoolTopology::Labelled(label) => {
+                        let mut matching_key = false;
+                        for key in label.affinity.keys() {
+                            if label.inclusion.contains_key(key) {
+                                matching_key = true;
+                            }
+                        }
+                        !matching_key && !label.affinity.is_empty()
+                    }
+                },
+                None => false,
+            };
+            node_affinity || pool_affinity
+        }
+        None => false,
+    }
+}
+
+/// Return a list of pre sorted pools in case of affinity.
+pub(crate) fn group_pools_by_affinity(
+    mut pools: Vec<PoolWrapper>,
+    suitable_pools: GetSuitablePools,
+    registry: &Registry,
+) -> Vec<Vec<PoolWrapper>> {
+    let mut map: HashMap<String, Vec<PoolWrapper>> = HashMap::new();
+    for pool in pools.iter_mut() {
+        let id = pool.state().id.clone();
+        let spec = registry.specs().pool(&id).unwrap_or_default();
+        let pool_labels = spec.labels.clone().unwrap_or_default();
+
+        // Get the affinity labels from the pool topology
+        let pool_affinity = match suitable_pools
+            .spec()
+            .topology
+            .as_ref()
+            .and_then(|topology| topology.pool.as_ref())
+        {
+            Some(PoolTopology::Labelled(label)) => label.affinity.clone(),
+            _ => HashMap::new(),
+        };
+
+        // performance: gold and zone: us-west, then the map key is "performance:gold;zone:us-west;"
+        // and the values are the vec of pool wrappers that have these labels
+        for (pool_affinity_key, _) in pool_affinity.iter() {
+            let mut result = String::new();
+            let mut labels = HashMap::new();
+            if let Some(value) = pool_labels.get(pool_affinity_key) {
+                result.push_str(pool_affinity_key);
+                result.push_str(":");
+                result.push_str(value);
+                result.push_str(";");
+                labels.insert(pool_affinity_key.clone(), value.clone());
+            }
+            pool.set_labels(Some(labels.clone()));
+            if let Some(pools) = map.get_mut(&result) {
+                pools.push(pool.clone());
+            } else {
+                map.insert(result, vec![pool.clone()]);
+            }
+        }
+    }
+
+    let mut final_list_of_pools: Vec<Vec<PoolWrapper>> = map.into_values().collect();
+
+    // Sort vectors by their lengths in descending order
+    final_list_of_pools.sort_by_key(|b| std::cmp::Reverse(b.len()));
```

Comment:
Please add a rust unittest for this logic (in another pr is fine, but do it as the next)